### PR TITLE
Implement AppDomain.Monitoring*MemorySize

### PR DESCRIPTION
**Pull request by stephentoub**
Tools like Benchmark.NET use this API to get total allocated memory.  Any reason not to make it work by just calling GC.GetTotalAllocatedBytes?

cc: @jkotas, @Maoni0, @VSadov, @luhenry

### CODE REVIEW

**Review by jkotas — APPROVED** [review]

File: src/System.Private.CoreLib/shared/System/AppDomain.cs [hunk 1]
```diff
@@ -199,7 +199,7 @@ public static bool MonitoringIsEnabled
 
         public static long MonitoringSurvivedProcessMemorySize { get { throw CreateResMonNotAvailException(); } }
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/AppDomain.cs]: You can also implement this one as `GC.GetGCMemoryInfo().HeapSizeBytes`

Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/AppDomain.cs]: Cool, was wondering about that, will do.

Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/AppDomain.cs]: Since there aren't multiple AppDomains, MonitoringSurvivedMemorySize could then also just call MonitoringSurvivedProcessMemorySize, right?

Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/AppDomain.cs]: Right

File: src/System.Private.CoreLib/shared/System/AppDomain.cs [hunk 2]
```diff
@@ -195,15 +195,13 @@ public static bool MonitoringIsEnabled
             }
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/AppDomain.cs]: I am wondering whether we should also make `MonitoringIsEnabled` return `true`, and delete `throw new PlatformNotSupportedException(SR.PlatformNotSupported_AppDomain_ResMon);` from the setter.

Then the only wrinkle that remains is `MonitoringTotalProcessorTime` that is "just work" to implement: using `GetProcessTimes` and `getrusage`. This can be separate PR.

**Comment by VSadov** [comment]
I think the only concern is the quirk where `GC.GetTotalAllocatedBytes` may be revised lower when threads die. I am not sure if `MonitoringTotalAllocatedMemorySize` had that. 
By briefly looking at the implementation - it did not have that (which means it would consider unspent bytes of dead threads as allocated, which is technically incorrect).

These is a fairly noisy/racy metric. My only worry, but not a huge one, is that some consumer might go crazy when seeing decreased allocations.

**Comment by VSadov** [comment]
Note that not subtracting the unspent bytes is kind of bad - it is not just an error, it is an error that could accumulate with more threads terminated.  I.E - it would not be "noise that will average out".

I wonder if we may tweak `GC.GetTotalAllocatedBytes` to remember the high watermark of reported value and not go below that, and stop worrying about some observer seeing decreasing values.

**Comment by VSadov** [comment]
The change looks good, but we should look at making `GC.GetTotalAllocatedBytes` results monotonic.

**Review by VSadov — APPROVED** [review]

**Comment by VSadov** [comment]
Logged https://github.com/dotnet/coreclr/issues/24615 - " Consider making `GC.GetTotalAllocatedBytes` monotonic"

**Comment by jkotas** [comment]
Failing CoreFX tests should be disabled in https://github.com/dotnet/coreclr/blob/master/tests/CoreFX/CoreFX.issues.rsp

**Review by jkotas — APPROVED** [review]
Nice!

File: src/System.Private.CoreLib/shared/System/AppDomain.cs (hunk 2 above)
Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/AppDomain.cs]: Good idea.  Done.

File: tests/CoreFX/CoreFX.issues.json [hunk 3]
```diff
@@ -1128,6 +1128,26 @@
             "namespaces": null,
             "classes": null,
             "methods": [
+                {
+                    "name": "System.Tests.AppDomainTests.MonitoringIsEnabled",
```
Comment (stephentoub) [inline tests/CoreFX/CoreFX.issues.json]: @jkotas, has the suppression mechanism changed?  These tests are still showing up at https://mc.dot.net/#/user/coreclr-ci/pr~2Fdotnet~2Fcoreclr~2Frefs~2Fpull~2F24610~2Fmerge/test~2Ffunctional~2Fcli~2F/20190520.764/workItem/System.Runtime.Extensions.Tests as failing, even though I think I've suppressed them here correctly.  Do you see anything I've done incorrectly?

Comment (jkotas) [reply tests/CoreFX/CoreFX.issues.json]: Yes, the exclusion file was changed recently. I have updated the docs: https://github.com/dotnet/coreclr/commit/d498abd2ec0937bbc5005d2d8d51add89c58a946 . I will submit PR to cleanup files that are no longer used.

Comment (stephentoub) [reply tests/CoreFX/CoreFX.issues.json]: Aha, thanks, I'll fix the PR.

Comment (jkotas) [reply tests/CoreFX/CoreFX.issues.json]: https://github.com/dotnet/coreclr/pull/24684

**Comment by Maoni0** [comment]
@VSadov 

> which means it would consider unspent bytes of dead threads as allocated, which is technically incorrect

this all depends on your definition of "allocated bytes", from GC's POV, that was what it gave out (meaning nothing else can use this memory anymore except whoever got this alloc context) so it was its view of what's allocated. 

but I don't have a problem with returning a different definition of allocated bytes which makes more sense for benchmark monitoring.

File: src/System.Private.CoreLib/shared/System/AppDomain.cs [hunk 4]
```diff
@@ -184,26 +184,21 @@ public static void Unload(AppDomain domain)
...
                 }
-                throw new PlatformNotSupportedException(SR.PlatformNotSupported_AppDomain_ResMon);
             }
         }
 
-        public long MonitoringSurvivedMemorySize { get { throw CreateResMonNotAvailException(); } }
+        public long MonitoringSurvivedMemorySize => MonitoringSurvivedProcessMemorySize;
 
-        public static long MonitoringSurvivedProcessMemorySize { get { throw CreateResMonNotAvailException(); } }
+        public static long MonitoringSurvivedProcessMemorySize => GC.GetGCMemoryInfo().HeapSizeBytes;
```
Comment (Maoni0) [inline src/System.Private.CoreLib/shared/System/AppDomain.cs]: >GC.GetGCMemoryInfo().HeapSizeBytes [](start = 66, length = 34)

HeapSize also includes Fragmentation so this should be (HeapSize - Fragmentation) which would be survived.

File: src/System.Private.CoreLib/shared/System/AppDomain.cs [hunk 5]
```diff
@@ -184,26 +184,21 @@ public static void Unload(AppDomain domain)
...
 
-        public long MonitoringTotalAllocatedMemorySize { get { throw CreateResMonNotAvailException(); } }
-
-        public TimeSpan MonitoringTotalProcessorTime { get { throw CreateResMonNotAvailException(); } }
-
-        private static Exception CreateResMonNotAvailException() => new InvalidOperationException(SR.PlatformNotSupported_AppDomain_ResMon);
+        public long MonitoringTotalAllocatedMemorySize => GC.GetTotalAllocatedBytes();
```
Comment (Maoni0) [inline src/System.Private.CoreLib/shared/System/AppDomain.cs]: >MonitoringTotalAllocatedMemorySize [](start = 20, length = 34)

I'm wondering if it makes sense for this to also give users an option for precious monitoring (which would call GC.GetTotalAllocatedBytes with true). I don't have a strong opinion on this - perhaps someone who might want to use this should comment - @adamsitnik?

Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/AppDomain.cs]: > I'm wondering if it makes sense for this to also give users an option for precious monitoring (which would call GC.GetTotalAllocatedBytes with true)

@Maoni0, which of `GetTotalAllocatedBytes(true)` or `GetTotalAllocatedBytes(false)` maps better to `MonitoringTotalAllocatedMemorySize` on netfx today?  We can just use whichever is closer, since the whole point of this is to enable existing such code to keep working.  Someone who wants the other semantics can just use `GetTotalAllocatedBytes` directly, or other such APIs we expose.

File: src/System.Private.CoreLib/shared/System/AppDomain.cs (hunk 4 above)
Comment (stephentoub) [reply src/System.Private.CoreLib/shared/System/AppDomain.cs]: Ok, will change, thanks.

File: src/System.Private.CoreLib/shared/System/AppDomain.cs (hunk 5 above)
Comment (adamsitnik) [reply src/System.Private.CoreLib/shared/System/AppDomain.cs]: FWIW As the end user of this API I would expect it to call `GC.GetTotalAllocatedBytes(true)`. The way I see it is that `MonitoringTotalAllocatedMemorySize` requires `AppDomain.MonitoringIsEnabled = true;` on Full .NET Framework which means that the monitoring has a cost.

Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/AppDomain.cs]: > Full .NET Framework which means that the monitoring has a cost.

This cost was in per-AppDomain accounting, e.g. enabling monitoring made the AppDomain switching more expensive.

These APIs have not provided accurate results in .NET Framework. `GC.GetTotalAllocatedBytes(/*false*/)` is a good approximation for what these APIs have done in .NET Framework.

File: src/System.Private.CoreLib/shared/System/AppDomain.cs [hunk 6]
```diff
@@ -184,26 +184,28 @@ public static void Unload(AppDomain domain)
...
+        {
+            get
+            {
+                GCMemoryInfo mi = GC.GetGCMemoryInfo();
+                return mi.HeapSizeBytes - mi.FragmentedBytes;
+            }
+        }
 
-        private static Exception CreateResMonNotAvailException() => new InvalidOperationException(SR.PlatformNotSupported_AppDomain_ResMon);
+        public long MonitoringTotalAllocatedMemorySize => GC.GetTotalAllocatedBytes(precise: true);
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/AppDomain.cs]: Imprecise result is a better functionality and performance match for what this API does on  .NET Framework.
